pkg/nuclide-rust/lib/BuckIntegration.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the license found in the LICENSE file in
+ * the root directory of this source tree.
+ *
+ * @flow
+ * @format
+ */
+
+import type {TaskInfo} from '../../nuclide-buck/lib/types';
+import type {
+  AtomLanguageService,
+  LanguageService,
+} from '../../nuclide-language-service';
+
+import {getLogger} from 'log4js';
+import fsPromise from 'nuclide-commons/fsPromise';
+import {
+  getRustInputs,
+  getSaveAnalysisTargets,
+  normalizeNameForBuckQuery,
+} from './BuckUtils';
+
+import * as BuckService from '../../nuclide-buck-rpc';
+
+const logger = getLogger('nuclide-rust');
+
+export async function updateRlsBuildForTask(
+  task: TaskInfo,
+  service: AtomLanguageService<LanguageService>,
+) {
+  const buildTarget = normalizeNameForBuckQuery(task.buildTarget);
+
+  // TODO: Filter by known Rust build targets
+  const files = await getRustInputs(task.buckRoot, buildTarget);
+  // Probably not a Rust build target, ignore
+  if (files.length === 0) {
+    return;
+  }
+  // We need only to pick a representative file to get a related lang service
+  const fileUri = task.buckRoot + '/' + files[0];
+  logger.debug(`fileUri: ${fileUri}`);
+
+  const langService = await service.getLanguageServiceForUri(fileUri);
```

Comment:
There's an invalid assumption here that the multi-project language service will only have one LSP server that we need to update its build command (while in reality there can be multiple buck roots, each having a different LSP server -- according to the project files config below).

Do you need to `getAllLanguageServices` or `observeLanguageServices`?
Can you test with multiple buck roots to see if things work properly?

Reply:
I understood that MultiLspLanguageService spawn separate LSP server per appropriate 'top-level' project config files like .hhconfig and that calling this seems like a good way to retrieve a handle to appropriate LSP server that's responsible for the fileUri we're building.

However, I did assume there's going to be a single Buck root; I'll see if it works with multiple ones.

Reply:
@mostafaeweda sorry it took so long! Just tested this with multiple buck roots, using https://github.com/Xanewok/rust-buck-skeleton and building both `//server:server` and `nested//:nested`. This correctly retrieved the LSP server at root dir and nested/, respectively, so this looks good so far =)